pkg/platform/abstract/platform.go

```diff
@@ -688,6 +702,108 @@ func (ap *Platform) GetProcessorLogsAndBriefError(scanner *bufio.Scanner) (strin
 	return common.FixEscapeChars(formattedProcessorLogs), common.FixEscapeChars(briefErrorsMessage)
 }
 
+func (ap *Platform) GetProjectFunctions(namespace, projectName string) ([]platform.Function, error) {
+	return ap.platform.GetFunctions(&platform.GetFunctionsOptions{
+		Namespace: namespace,
+		Labels:    fmt.Sprintf("nuclio.io/project-name=%s", projectName),
+	})
+}
+
+func (ap *Platform) GetProjectAPIGateways(namespace, projectName string) ([]platform.APIGateway, error) {
+	apiGateways, err := ap.platform.GetAPIGateways(&platform.GetAPIGatewaysOptions{
+		Namespace: namespace,
+		Labels:    fmt.Sprintf("nuclio.io/project-name=%s", projectName),
+	})
+	if err != nil {
+
+		// if api gateways are not supported on this platform, ignore
+		if err == platform.ErrUnsupportedMethod {
+			return nil, nil
+		}
+
+		return nil, errors.Wrap(err, "Failed to get api gateways")
+	}
+	return apiGateways, nil
+}
+
+func (ap *Platform) DeleteProjectResources(namespace string, projectName string, waitForDeletionCompletion bool) error {
+	ap.Logger.InfoWith("Deleting project resources",
+		"projectName", projectName,
+		"namespace", namespace)
+
+	doneChan := make(chan error, 1)
+	go func() {
+		if err := ap.deleteProjectResources(namespace, projectName); err != nil {
+
+			// no one waiting, print to log
+			if !waitForDeletionCompletion {
+				ap.Logger.WarnWith("Failed to delete project related resources",
+					"err", errors.GetErrorStackString(err, 10))
+			}
+
+			doneChan <- errors.Wrap(err, "Failed to delete project resources")
+		}
+		doneChan <- nil
+	}()
+
+	// For testing and clients that requires it (nuctl)
+	if waitForDeletionCompletion {
+		return <-doneChan
+	}
+
+	return nil
+}
+
+func (ap *Platform) deleteProjectResources(namespace, projectName string) error {
+	// NOTE: functions delete their related function events
+
+	// get functions
+	functions, err := ap.GetProjectFunctions(namespace, projectName)
+	if err != nil {
+		return errors.Wrap(err, "Failed to get project functions")
+	}
+
+	// get api gateways
+	apiGateways, err := ap.GetProjectAPIGateways(namespace, projectName)
+	if err != nil {
+		return errors.Wrap(err, "Failed to get project api gateways")
+	}
+
+	deleteAPIGatewaysErrGroup, _ := errgroup.WithContext(context.TODO())
+
+	// delete api gateways
+	for _, apiGateway := range apiGateways {
+		apiGateway := apiGateway
+		deleteAPIGatewaysErrGroup.Go(func() error {
+			if err := ap.platform.DeleteAPIGateway(&platform.DeleteAPIGatewayOptions{
+				Meta: apiGateway.GetConfig().Meta,
+			}); err != nil {
+				return errors.Wrapf(err, "Failed to delete api gateway '%s'", apiGateway.GetConfig().Meta.Name)
+			}
+			return nil
+		})
+	}
+	if err := deleteAPIGatewaysErrGroup.Wait(); err != nil {
+		return errors.Wrap(err, "Failed deleting project api gateways")
+	}
```

Comment:
Not sure needed in this PR, but the right way is to leverage k8s's `DeleteCollection` and not doing tons of calls to `Delete`

Reply:
That is a good optimization indeed. I will add a comment.

But do know that deleteCollection requires label selector / field selector and not by a given list of names.